docs/tutorials/stateful-application/zookeeper.md

```diff
@@ -59,51 +57,43 @@ After this tutorial, you will know the following.
 
 ### ZooKeeper Basics
 
-[Apache ZooKeeper](https://zookeeper.apache.org/doc/current/) is a 
+[Apache ZooKeeper](https://zookeeper.apache.org/doc/current/) is a
 distributed, open-source coordination service for distributed applications.
-ZooKeeper allows you to read, write, and observe updates to data. Data are 
-organized in a file system like hierarchy and replicated to all ZooKeeper 
-servers in the ensemble (a set of ZooKeeper servers). All operations on data 
-are atomic and sequentially consistent. ZooKeeper ensures this by using the 
-[Zab](https://pdfs.semanticscholar.org/b02c/6b00bd5dbdbd951fddb00b906c82fa80f0b3.pdf) 
+ZooKeeper allows you to read, write, and observe updates to data. Data are
+organized in a file system like hierarchy and replicated to all ZooKeeper
+servers in the ensemble (a set of ZooKeeper servers). All operations on data
+are atomic and sequentially consistent. ZooKeeper ensures this by using the
+[Zab](https://pdfs.semanticscholar.org/b02c/6b00bd5dbdbd951fddb00b906c82fa80f0b3.pdf)
 consensus protocol to replicate a state machine across all servers in the ensemble.
 
-The ensemble uses the Zab protocol to elect a leader, and
-data can not be written until a leader is elected. Once a leader is 
-elected, the ensemble uses Zab to ensure that all writes are replicated to a 
-quorum before they are acknowledged and made visible to clients. Without respect
-to weighted quorums, a quorum is a majority component of the ensemble containing 
-the current leader. For instance, if the ensemble has three servers, a component 
-that contains the leader and one other server constitutes a quorum. If the 
-ensemble can not achieve a quorum, data can not be written.
-
-ZooKeeper servers keep their entire state machine in memory, but every mutation 
-is written to a durable WAL (Write Ahead Log) on storage media. When a server 
-crashes, it can recover its previous state by replaying the WAL. In order to 
-prevent the WAL from growing without bound, ZooKeeper servers will periodically 
-snapshot their in memory state to storage media. These snapshots can be loaded 
-directly into memory, and all WAL entries that preceded the snapshot may be 
-safely discarded.
+The ensemble uses the Zab protocol to elect a leader, and the ensemble cannot write data until it elects a leader. Once it elects a leader, the ensemble uses Zab to ensure that it replicates all writes to a
+quorum before it acknowledges and makes them visible to clients. Without respect
+to weighted quorums, a quorum is a majority component of the ensemble containing
+the current leader. For instance, if the ensemble has three servers, a component
+that contains the leader and one other server constitutes a quorum. If the
+ensemble can not achieve a quorum, the ensemble data can not write data.
```

Comment:
"the ensemble data can not write data" -- typo? Think again about passive voice -- seems clearer here.
s/can not/cannot (unless the meaning really is "might not be" as opposed to "won't be")